File: airbyte-commons/src/main/java/io/airbyte/commons/features/FeatureFlagHelper.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.features;

import com.google.common.annotations.VisibleForTesting;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class FeatureFlagHelper {

  public static boolean isFieldSelectionEnabledForWorkspace(final FeatureFlags featureFlags, final UUID workspaceId) {
    return isWorkspaceIncludedInFlag(featureFlags, FeatureFlags::fieldSelectionWorkspaces, workspaceId, "field selection")
        || featureFlags.applyFieldSelection();
  }

  @VisibleForTesting
  static boolean isWorkspaceIncludedInFlag(final FeatureFlags featureFlags,
                                           final Function<FeatureFlags, String> flagRetriever,
                                           final UUID workspaceId,
                                           final String context) {
    final String workspaceIdsString = flagRetriever.apply(featureFlags);
    final Set<UUID> workspaceIds = new HashSet<>();
    if (workspaceIdsString != null && !workspaceIdsString.isEmpty()) {
      for (final String id : workspaceIdsString.split(",")) {
        try {
          workspaceIds.add(UUID.fromString(id));
        } catch (final IllegalArgumentException e) {
          log.warn("Malformed workspace id for {}: {}", context, id);
        }
      }
    }
    return workspaceId != null && workspaceIds.contains(workspaceId);
  }

}

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/features/FeatureFlagHelper.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.commons.features

import com.google.common.annotations.VisibleForTesting
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.*
import java.util.function.Function

private val log = KotlinLogging.logger {}

object FeatureFlagHelper {
    fun isFieldSelectionEnabledForWorkspace(
        featureFlags: FeatureFlags,
        workspaceId: UUID?
    ): Boolean {
        return (isWorkspaceIncludedInFlag(
            featureFlags,
            { obj: FeatureFlags -> obj.fieldSelectionWorkspaces() },
            workspaceId,
            "field selection"
        ) || featureFlags.applyFieldSelection())
    }

    @VisibleForTesting
    fun isWorkspaceIncludedInFlag(
        featureFlags: FeatureFlags,
        flagRetriever: Function<FeatureFlags, String?>,
        workspaceId: UUID?,
        context: String?
    ): Boolean {
        val workspaceIdsString = flagRetriever.apply(featureFlags)
        val workspaceIds: MutableSet<UUID> = HashSet()
        if (!workspaceIdsString.isNullOrEmpty()) {
            for (id in
                workspaceIdsString
                    .split(",".toRegex())
                    .dropLastWhile { it.isEmpty() }
                    .toTypedArray()) {
                try {
                    workspaceIds.add(UUID.fromString(id))
                } catch (e: IllegalArgumentException) {
                    log.warn { "Malformed workspace id for $context: $id" }
                }
            }
        }
        return workspaceId != null && workspaceIds.contains(workspaceId)
    }
}
